client/mm/exchange_adaptor.go

```diff
@@ -330,44 +347,142 @@ func (u *unifiedExchangeAdaptor) SufficientBalanceForCEXTrade(baseID, quoteID ui
+	var baseDelta, quoteDelta int64
+	if trade.Sell {
+		baseDelta = -int64(trade.BaseFilled)
+		quoteDelta = int64(trade.QuoteFilled)
+	} else {
+		baseDelta = int64(trade.BaseFilled)
+		quoteDelta = -int64(trade.QuoteFilled)
+	}
+
+	e := &MarketMakingEvent{
+		ID:         eventID,
+		TimeStamp:  timestamp,
+		Pending:    !trade.Complete,
+		BaseDelta:  baseDelta,
+		QuoteDelta: quoteDelta,
+		CEXOrderEvent: &cexOrderEvent{
+			ID:          trade.ID,
+			Rate:        trade.Rate,
+			Qty:         trade.Qty,
+			Sell:        trade.Sell,
+			BaseFilled:  trade.BaseFilled,
+			QuoteFilled: trade.QuoteFilled,
+		},
+	}
 
-		locked := map[uint32]uint64{}
-		locked[fromAsset] += o.LockedAmt
-		locked[fromFeeAsset] += o.ParentAssetLockedAmt + o.RefundLockedAmt
-		locked[toFeeAsset] += o.RedeemLockedAmt
+	u.eventLogDB.storeEvent(u.startTime.Load(), u.market, e)
+	u.notifyEvent(e)
+}
 
-		orderInfo := idToInfo[orderID]
-		u.pendingDEXOrders[orderID] = &pendingDEXOrder{
-			swaps:            make(map[string]*asset.WalletTransaction),
-			redeems:          make(map[string]*asset.WalletTransaction),
-			refunds:          make(map[string]*asset.WalletTransaction),
-			availableDiff:    availableDiff,
-			locked:           locked,
-			pending:          map[uint32]uint64{},
-			order:            o,
-			placementIndex:   orderInfo.placementIndex,
-			counterTradeRate: orderInfo.counterTradeRate,
+// updateDepositEvent updates the event log with the current state of a
+// pending deposit and sends an event notification.
+func (u *unifiedExchangeAdaptor) updateDepositEvent(eventID uint64, timestamp int64, assetID uint32, tx *asset.WalletTransaction, amtCredited uint64, complete bool) {
+	var baseDelta, quoteDelta, diff int64
+	var baseFees, quoteFees uint64
+	if complete && tx.Amount > amtCredited {
+		diff = int64(tx.Amount - amtCredited)
+	}
+	if assetID == u.market.BaseID {
+		baseDelta = -diff
+		baseFees = tx.Fees
+	} else {
+		quoteDelta = -diff
+		quoteFees = tx.Fees
+	}
+	e := &MarketMakingEvent{
+		ID:         eventID,
+		TimeStamp:  timestamp,
+		BaseDelta:  baseDelta,
+		QuoteDelta: quoteDelta,
+		BaseFees:   baseFees,
+		QuoteFees:  quoteFees,
+		Pending:    !complete,
+		DepositEvent: &depositEvent{
+			AssetID:     assetID,
+			Transaction: tx,
+			CEXCredit:   amtCredited,
+		},
+	}
+	u.eventLogDB.storeEvent(u.startTime.Load(), u.market, e)
+	u.notifyEvent(e)
+}
+
+// updateWithdrawalEvent updates the event log with the current state of a
+// pending withdrawal and sends an event notification.
+func (u *unifiedExchangeAdaptor) updateWithdrawalEvent(withdrawalID string, withdrawal *pendingWithdrawal, tx *asset.WalletTransaction) {
+	var baseFees, quoteFees uint64
+	complete := tx != nil && tx.Confirmed
+	if complete && withdrawal.amtWithdrawn > tx.Amount {
+		if withdrawal.assetID == u.market.BaseID {
+			baseFees = withdrawal.amtWithdrawn - tx.Amount
+		} else {
+			quoteFees = withdrawal.amtWithdrawn - tx.Amount
 		}
 	}
```

Comment:
Is there an assumption here that `updateWithdrawalEvent` will only be called once with `Confirmed = true`?

Reply:
It is called twice, once when the withdrawal is initiated and once when it is complete.